lighthouse/mcs-compliance.md

```diff
@@ -0,0 +1,186 @@
+# Add support for 100% MCS compliance
+
+Related issue:
+[Add support for 100% MCS Compliance](https://github.com/submariner-io/enhancements/issues/229)
+
+## Summary
+
+Current Lighthouse implementation differs from MCS API [^1] on two key points for ClusterIP Services:
+
+1. There is no VirtualIP [VIP] assigned in `ServiceImport`.
+2. `EndpointSlices` do not contain individual PodIPs.
+
+To better align with MCS API Lighthouse needs an option to allocate and track VirtualIPs in `ServiceImports` and use
+PodIPs in `EndpointSlices`.
+
+## Proposal
+
+When creating `ServiceExport`, users will have the option to add an annotation for enabling VIP for that Service.
+This annotation will also be added to the aggregated `ServiceImport`.
+
+A configuration flag will also be added to the `Submariner` and `ServiceDiscovery` CRDs, and `subctl` to set the
+options at the global level. This is for deployments where the user wants the option set for all exported Services and
+doesn't want to set the annotation explicitly for each `ServiceExport`.
+
+Default behavior with nothing configured would be same as it is currently i.e. no VIP and no PodIPs.
+
+Global flag can't be changed at runtime and requires a reinstall of whole clusterset. Change of flag in CRs will require
+a restart of Submariner on that cluster as it changes the behavior of `ServiceExports` without annotation.
+
+### VIP CIDR and allocation
+
+1. Default CIDR of `243.0.0/8` will be used for Clusterset. User can change this at the time of deploying broker.
+2. Each Cluster will get a default subset CIDR of `243.0.0.0/16` for allocating VIPs to Services exported from that
+cluster. User can change this at the time of join but this CIDR must be an unallocated subset of global VIP CIDR.
+3. VIP allocation is done by the first cluster to export the `Service`.
+4. When Service is exported on subsequent clusters, no VIP allocation is done if VIP already present on `ServiceImport`.
+5. VIP is deallocated only when `ServiceImport` is deleted i.e. `Service` or `ServiceExport` are no longer present on any
+of the clusters.
+6. If Submariner is uninstalled on the cluster that allocated a given VIP, VIP is not changed on `ServiceImport`.
```

Comment:
Some corner case Qs:

Regarding 
`If Submariner is uninstalled on the cluster that allocated a given VIP, VIP is not changed on ServiceImport`
Don't we delete all serviceimport CRs when submariner is uninstalled ?

Do we expect that cluster will get the same VIPs cidr subset after reinstall ?

Regarding 
`VIP allocation is done by the first cluster to export the Service.` 
do we need to implement  some lock mechanism to resolve race condition in case two clusters export service at the same time ?

Reply:
No - that’s handled by using K8s retry on conflict functionality.

Reply:
> Some corner case Qs:
> 
> Regarding `If Submariner is uninstalled on the cluster that allocated a given VIP, VIP is not changed on ServiceImport` Don't we delete all serviceimport CRs when submariner is uninstalled ?
Aggregated ServiceImport will still exist if service is exported on other clusters. Deletion will only happen if this was the last cluster to export service, in which case it will be deleted. Point here is that if the cluster allocating VIP is no longer there [deleted or submariner removed] we don't try to change the VIP as it will impact existing clients.
> 
> Do we expect that cluster will get the same VIPs cidr subset after reinstall ?
No. We don't make any assumptions and it is also possible that some other cluster gets that CIDR. We can see this happen with Globalnet.
> 
> Regarding `VIP allocation is done by the first cluster to export the Service.` do we need to implement some lock mechanism to resolve race condition in case two clusters export service at the same time ?
Not needed. As both try to create Aggregated ServiceImport, only one will be able to do so. Other will get conflict with `Resource already exists` error.